xds/src/main/java/io/grpc/xds/XdsNameResolver2.java

```diff
@@ -0,0 +1,391 @@
+        if (action.getCluster() != null) {
+          clusters.add(action.getCluster());
+        } else if (action.getWeightedCluster() != null) {
+          for (ClusterWeight weighedCluster : action.getWeightedCluster()) {
+            clusters.add(weighedCluster.getName());
+          }
+        }
+      }
+      boolean receivedNewCluster = false;
+      for (String newCluster : clusters) {
+        if (clusterRefs.putIfAbsent(newCluster, new AtomicInteger(1)) == null) {
+          receivedNewCluster = true;
+        }
+      }
+      // Update service config to include newly added clusters.
+      if (receivedNewCluster) {
+        updateResolutionResult();
+      }
+      // Make newly added clusters selectable by config selector.
+      routes = update.getRoutes();
+      // Drops reference for deleted clusters, update service config to remove deleted clusters
+      // not in use.
+      boolean shouldUpdateResult = false;
+      for (Map.Entry<String, AtomicInteger> entry : clusterRefs.entrySet()) {
+        if (!clusters.contains(entry.getKey())) {
+          int count = entry.getValue().decrementAndGet();
+          if (count == 0) {
+            clusterRefs.remove(entry.getKey());
+            shouldUpdateResult = true;
+          }
+        }
+      }
+      if (shouldUpdateResult) {
+        updateResolutionResult();
+      }
+    }
+
+    @Override
+    public void onResourceDoesNotExist(String resourceName) {
+      logger.log(XdsLogLevel.INFO, "Resource {0} is unavailable", resourceName);
+      ConfigOrError parsedServiceConfig =
+          serviceConfigParser.parseServiceConfig(Collections.<String, Object>emptyMap());
+      ResolutionResult result =
+          ResolutionResult.newBuilder()
+              .setServiceConfig(parsedServiceConfig)
+              // let channel take action for no config selector
+              .build();
+      listener.onResult(result);
+    }
+
+    @Override
+    public void onError(Status error) {
+      logger.log(
+          XdsLogLevel.WARNING,
+          "Received error from xDS client {0}: {1}", xdsClient, error.getDescription());
+      listener.onError(error);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private ConfigOrError parseServiceConfig(String serviceConfigJson) {
+    Map<String, ?> serviceConfig;
+    try {
+      serviceConfig = (Map<String, ?>) JsonParser.parse(serviceConfigJson);
+    } catch (IOException e) {
+      return ConfigOrError.fromError(
+          Status.INTERNAL.withCause(e).withDescription("bug: malformed service config"));
+    }
+    return serviceConfigParser.parseServiceConfig(serviceConfig);
+  }
+
+  @VisibleForTesting
+  static String generateServiceConfigWithMethodConfig(
+      String fullMethodName, long timeoutNano) {
+    int index = fullMethodName.lastIndexOf('/');
+    String serviceName = fullMethodName.substring(0, index);
+    String methodName = fullMethodName.substring(index + 1);
+    StringBuilder sb = new StringBuilder();
```

Comment:
Yeah, previously we used to generate the data structure, I mean parsed JSON (aka, `Map<String, ?>`). But that was hard to read given that most times the config is deeply nested and reviewers were usually having hard time (from my personal experience) following the logic of maps of maps of maps, etc. If you are talking about directly generate the parsed service config (aka, `ManagedChannelServiceConfig`), isn't that kept as channel's internal on purpose so that plugins do not depend on that API?

Reply:
As we discussed offline, switched to generating `Map<String, ?>` instead of doing string concatenation. I still keep the way of using JsonParser and JSON string for doing testing, as that is much easier for verifying the output and read/maintain the test.